File: book_kit/src/main/java/net/lzbook/kit/data/bean/Book.java
package net.lzbook.kit.data.bean;

import android.text.TextUtils;

import java.io.Serializable;

public class Book implements Serializable, Comparable<Book>, Cloneable {
    //0:线上书籍
    public static final int TYPE_ONLINE = 0;
    private static final long serialVersionUID = -6628109520747996395L;
    //区分小说在书架上的位置是否要放置广告
    public int book_type;
    //原先小说组ID，保留用于替换book_id
    public int gid;
    //小说id
    public int nid;
    //记录书籍当前的阅读位置
    public int sequence = -2;
    //书签偏移量
    public int offset = -1;
    //定义的章节序号的最新序号
    public int last_sort;
    //最佳目录的章节序号
    public int gsort = -1;
    //更新状态
    public int update_status = -1;
    //获取小说数据失败的信息
    public String bad_nid;
    //添加到书架记录当前sequence的时间
    public long sequence_time;
    //是否添加书签
    public int readed;
    //初始化书籍信息时，加入到数据库的时间
    public long insert_time;
    //升级后使用的字段
    //小说名称
    public String name;
    //小说作者名
    public String author;
    //小说简介
    public String desc;
    //小说分类信息
    public String category;
    //小说封面页图片URL
    public String img_url;
    //判断小说当前更新状态
    public int status = -1;
    public String site;
    //小说最新章节名称
    public String last_chapter_name;
    //小说章节数量
    public int chapter_count;
    //本地保存最新更新时间
    public long last_updatetime_native;

    //新的字段
    //每本小说都有其唯一的book_id
    public String book_id;
    //小说来源id
    public String book_source_id;
    //最新章节对应MD5值
    public String last_chapter_md5;
    //最新章节对应的url
    public String last_chapter_url;
    public String last_chapter_url1;

    //中间变量(用于存储搜狗md，百度bookID等字段)
    public String parameter;
    //预留字段
    public String extra_parameter;
    //升级初始化状态：0:暂不更新来源，1:来源更新，2:小说可切换到WEB，3:小说可切换到源站，4:小说下线; 5:小说弹目录
    public int initialization_status = 0;

    //用来标识书籍是最近阅读的
    public boolean isLastRead = false;

    public int dex = -1;
    //最后检查更新的时间
    public long last_checkupdatetime = 0;
    //最后最后一次本地更新成功的时间, 与最后一章节中的时间有区别
    public long last_updateSucessTime;

    //上一次更新到的章节序号，为了和青果适配而新增的字段
    public int chapters_update_index;
    public String readPersonNum;

    public int list_version = -1;
    public int c_version = -1;



    public int item_type = 0;
    public int item_position = -1;


    public static boolean isOnlineType(int type) {
        return type == TYPE_ONLINE;
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        } else if (!(o instanceof Book)) {
            return false;
        } else {
            Book book = (Book) o;
            if (!TextUtils.isEmpty(book.book_id)) {
                return book.book_id.equals(this.book_id);// 只要book_id相同就认为是同一个对象
            }
            return false;
        }

    }

    @Override
    public int hashCode() {
        return gid;
    }

    @Override
    public int compareTo(Book another) {
        return this.sequence_time == another.sequence_time ? 0 : (this.sequence_time < another.sequence_time ? 1 : -1);
    }

    @Override
    public Book clone() throws CloneNotSupportedException {
        return (Book) super.clone();
    }

    @Override
    public String toString() {
        return "Book {" +
                " book_id : " + book_id +
                " book_source_id : " + book_source_id +
                " name : " + name +
                " author : " + author +
                " parameter : " + parameter +
                " extra_parameter : " + extra_parameter +
                " site : " + site +
                " chapter_count: " + chapter_count +
                " sequence: " + sequence +
                " }";
    }

}

File: basic/src/main/java/com/ding/basic/bean/Book.kt
package com.ding.basic.bean

import android.arch.persistence.room.*
import android.view.ViewGroup
import com.ding.basic.db.migration.*
import java.io.Serializable

@Entity(tableName = "book")
open class Book : Serializable, Comparable<Book>, Cloneable {
    /***
     * 数据升级，字段问题
     * 1、字段冲突 -> book_type
     * 原来标识书籍种类，区分线上书籍或者广告类型
     * 现在标识书籍来源，区分智能书籍、青果书籍和VIP书籍
     * 2、类型冲突 -> state
     * 原来使用Int类型标识书籍完结、连载状态
     * 现在使用String类型标识书籍完结、连载状态
     *
     * 解决方法：
     * 1、修改原有book_type为item_type
     * 2、数据库迁移过程中手动修改
     */

    /************************************ 数据融合 ************************************/

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id: Int = 0

    //书籍ID
    @ColumnInfo(name = "book_id")
    var book_id: String = ""

    //书籍ChapterID，新版接口主要字段
    @ColumnInfo(name = "book_chapter_id")
    var book_chapter_id: String = ""

    //书籍SourceID，逐步废弃
    @ColumnInfo(name = "book_source_id")
    var book_source_id: String = ""

    //书籍名称
    @ColumnInfo(name = "name")
    var name: String? = null

    //书籍作者
    @ColumnInfo(name = "author")
    var author: String? = null

    //书籍描述
    @ColumnInfo(name = "desc")
    var desc: String? = null

    //书籍标签，字符串格式，逗号隔开
    @ColumnInfo(name = "label")
    var label: String? = null

    //一级分类
    @ColumnInfo(name = "genre")
    var genre: String? = null

    //二级分类
    @ColumnInfo(name = "sub_genre")
    var sub_genre: String? = null

    //书籍封面地址
    @ColumnInfo(name = "img_url")
    var img_url: String? = null

    //书籍状态：FINISH
    @FieldMigration(oldName = "status", converter = StatusConverter::class)
    @ColumnInfo(name = "status")
    var status: String? = null

    //书籍源站
    @FieldMigration(oldName = "site", converter = BookHostConverter::class)
    @ColumnInfo(name = "host")
    var host: String? = null

    @FieldMigration(oldName = "site", converter = BookTypeConverter::class)
    //书籍种类：zn智能；vip付费；qg青果
    @ColumnInfo(name = "book_type")
    var book_type: String? = null

    //书籍字数
    @ColumnInfo(name = "word_count")
    var word_count: String? = null

    //目录Version
    @ColumnInfo(name = "list_version")
    var list_version: Int = -1

    //内容Version
    @ColumnInfo(name = "c_version")
    var c_version: Int= -1

    //最新章节
    @FieldMigration(oldName = "last_chapter_name", converter = BookChapterConverter::class)
    @Embedded(prefix = "chapter_")
    var last_chapter: Chapter? = null


    @ColumnInfo(name= "score")
    var score: Float = 0.0f

    @ColumnInfo(name= "uv")
    var uv: Long = 0

    /************************************ 本地字段 ************************************/
    //书籍阅读状态
    @ColumnInfo(name = "readed")
    var readed: Int = 0

    //书籍阅读进度
    @ColumnInfo(name = "sequence")
    var sequence: Int = -2

    //阅读章节偏移量
    @ColumnInfo(name = "offset")
    var offset: Int = -1

    //书籍阅读时间
    @FieldMigration(oldName = "sequence_time")
    @ColumnInfo(name = "last_read_time")
    var last_read_time: Long = 0

    //书籍加入书架时间
    @ColumnInfo(name = "insert_time")
    var insert_time: Long = 0

    //书籍更新状态
    @ColumnInfo(name = "update_status")
    var update_status: Int = -1

    //小说章节数量
    @ColumnInfo(name = "chapter_count")
    var chapter_count: Int = 0

    //最后检查更新的时间,旧字段：last_updatetime_native
    @ColumnInfo(name = "last_check_update_time")
    var last_check_update_time: Long = 0

    //最后最后一次本地更新成功的时间, 与最后一章节中的时间有区别
    @ColumnInfo(name = "last_updateSucessTime")
    var last_update_success_time: Long = 0

    //数据融合升级，默认为0
    @ColumnInfo(name = "update_date_fusion")
    var update_date_fusion: Int = 0


    //上一次更新到的章节序号，为了和青果适配而新增的字段
    @ColumnInfo(name = "chapters_update_index")
    var chapters_update_index: Int = 0

    @ColumnInfo(name = "list_version_fix")
    var list_version_fix: Int = -1

    @ColumnInfo(name = "force_fix")
    var force_fix: Int = 0

    //区分书籍状态，目前没有本地书籍，暂时标识书籍和广告两种状态
    @Ignore
    var item_type: Int = 0
    //广告布局
    @Ignore
    @Transient
    var item_view: ViewGroup? = null

    //统计打点 当前页面是来自书籍封面/书架/上一页翻页
    @Ignore
    var fromType: Int = 0
    //统计当前阅读的小说是来自青果还是智能 打点统计
    @Ignore
    var channel_code: Int = 0


    constructor() : this(0)

    @Ignore
    constructor(id: Int) {
        this.id = id
    }

    /**
     * 书籍是否来源于青果
     */
    fun fromQingoo(): Boolean {
        host?.let {
            return it.contains("qingoo")
        }
        return false
    }

    /**
     * 书籍是否处于等待目录修复状态
     */
    fun waitingCataFix(): Boolean {
        if (force_fix == 1 || list_version < list_version_fix) {
            return true
        }
        return false
    }

    public override fun clone(): Book {
        return super.clone() as Book
    }

    override fun compareTo(other: Book): Int {
        return if (this.last_read_time == other.last_read_time) 0 else if (this.last_read_time < other.last_read_time) 1 else -1
    }

    override fun toString(): String {
        return "Book(id=$id, book_id='$book_id', book_chapter_id='$book_chapter_id', book_source_id='$book_source_id', name=$name, author=$author, desc=$desc, label=$label, genre=$genre, sub_genre=$sub_genre, img_url=$img_url, status=$status, host=$host, book_type=$book_type, word_count=$word_count, list_version=$list_version, c_version=$c_version, last_chapter=$last_chapter, readed=$readed, sequence=$sequence, offset=$offset, last_read_time=$last_read_time, insert_time=$insert_time, update_status=$update_status, chapter_count=$chapter_count, last_check_update_time=$last_check_update_time, last_update_success_time=$last_update_success_time, update_date_fusion=$update_date_fusion, chapters_update_index=$chapters_update_index, item_type=$item_type, fromType=$fromType, channel_code=$channel_code)"
    }

    companion object {
        const val STATUS_FINISH = "FINISH"
        const val STATUS_SERIALIZE = "SERIALIZE"
    }


}